libs/remote_storage/src/lib.rs

```diff
@@ -42,57 +39,96 @@ pub const DEFAULT_REMOTE_STORAGE_MAX_SYNC_ERRORS: u32 = 10;
 /// https://aws.amazon.com/premiumsupport/knowledge-center/s3-request-limit-avoid-throttling/
 pub const DEFAULT_REMOTE_STORAGE_S3_CONCURRENCY_LIMIT: usize = 100;
 
+#[derive(Clone, PartialEq)]
+pub struct RemoteObjectId(String);
+
+const S3_PREFIX_SEPARATOR: char = '/';
+
+///
+/// A key that refers to an object in remote storage. It works much like a Path,
+/// but it's a separate datatype so that you don't accidentally mix local paths
+/// and remote keys.
+///
+impl RemoteObjectId {
+    // Needed to retrieve last component for RemoteObjectId.
+    // In other words a file name
+    /// Turn a/b/c or a/b/c/ into c
+    pub fn object_name(&self) -> Option<&str> {
+        // corner case, char::to_string is not const, thats why this is more verbose than it needs to be
+        // see https://github.com/rust-lang/rust/issues/88674
+        if self.0.len() == 1 && self.0.chars().next().unwrap() == S3_PREFIX_SEPARATOR {
```

Comment:
Hmm. I think our disagreements here stem from different views on what RemoteObjectId represents. In my mind, RemoteObjectId is a "name" or "key" that can be used to refer to a remote object. It's not just an internal matter of the RemoteStorage implementation, it's something that the users of RemoteStorage need to construct and deal with. For example, to upload a file, you construct a RemoteObjectId, and then you upload the file with that ID. And later you can retrieve it back by using the same ID.

I think you're thinking of it as more of a physical locator in the remote storage. For the LocalFs implementation, it's the absolute path in the local filesystem to the file, and for S3 it's the prefix+key in S3. Am I right?

To me, it would feel more natural if RemoteObjectId only contained the caller-supplied part. For example, if you upload a file as `tenant1/timeline2/index.part` with the LocalFs implementation, RemoteObjectId should wrap just `tenant1/timeline2/index.part`, not `/data/storage/tenant1/timeline2/index.part`. The implementation will join the storage root and the RemoteObjectId when it opens the file for reading/writing. Similarly for S3, the bucket or prefix should not be part of RemoteObjectId.

That way, a RemoteObjectId would not be tied to a particular RemoteStorage implementation, or even to a particular RemoteStorage object. For example, if you wanted to copy a file from one RemoteStorage to another, you could do something like this:
```
let remote_object_id = RemoteObjectId("foo/bar");
remote_storage_src.download(remote_object_id);
remote_storage_dst.upload(remote_object_id);
```

Thoughts?

(Yeah, the code on RemoteObjectId should not involve S3_PREFIX_SEPARATOR.)

Reply:
Yes, I think you've got the idea right and I like this new proposal too, as something I've tried to describe [here](https://github.com/neondatabase/neon/pull/2387#issuecomment-1236876167).

My idea was to still keep the separation though, since the object pointers are different for every storage, but to keep some sort of a generic key, that could be converted into the "specific" key later by the repo itself.

Reply:
> My idea was to still keep the separation though, since the object pointers are different for every storage, but to keep some sort of a generic key, that could be converted into the "specific" key later by the repo itself.

I'm cool with that